### DISCUSSION TITLE
Transfer two local variables of MasterApp to one global variable of SubApp

### DISCUSSION THREAD

**Discussion by js-jixu** [idaholab/moose#22155] [Q&A General] [answered]
Hi, MOOSE developers.

I want to couple two input files by MultiApps & Transfers system. In MasterApp, there are two blocks, the temperature of the fluid block is `Tf`, and the temperature of the solid block is `Ts`. In the SubApp, the temperature variable for both the fluid and solid regions is defined as `T`. MasterApp and SubApp share one mesh file.

I calculated Tf and Ts in MasterApp, and now I want to transfer them to T of the corresponding block. What I mean is to transfer Ts to T of solid block, and transfer Tf to T of fluid block.

This is my ideal situation and results:

![1](https://user-images.githubusercontent.com/99813361/191546007-10df917c-6dcc-45e3-9b01-195febd6366c.png)

When I use the following syntax, the value of Ts in the fluid block (actually 0, because Ts is not defined in the fluid block in MasterApp) will overwrite the T of fluid block in SubApp passed by Tf.

```
[Transfers]
  [./to_sub_Tf]
    type = MultiAppProjectionTransfer
    direction = to_multiapp
    multi_app = SubApp
    source_variable = Tf
    variable = T
  [../]
  [./to_sub_Ts]
    type = MultiAppProjectionTransfer
    direction = to_multiapp
    multi_app = SubApp
    source_variable = Ts
    variable = T
  [../]
[]
```

This is the actual situation and results:

![2](https://user-images.githubusercontent.com/99813361/191546943-12c4a2e8-a929-48cd-a43e-4591b63956fb.png)

![3](https://user-images.githubusercontent.com/99813361/191546227-c1fab71c-bcde-4a9c-a613-0f671996e438.png)

How can I make T get the Ts and Tf of the corresponding block?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Basically you need to block-restrict your transfers. The last transfer is overwriting the previous value of T here. 

A few options:
- use a different transfer that supports block restriction
- implement block restriction (not hard) in your transfer
- use the transfers in this PR, which have block restriction https://github.com/idaholab/moose/pull/17417
- merge Ts and Tf in T_merged before transfering. You can use the SelfAux to do that or the ParsedAux

Guillaume

**Reply by js-jixu** [reply]
Thanks for your useful suggestions. I solved this problem based on your fourth suggestion. I defined an auxvariable named `T_merged`, then used `SelfAux` to merge Ts and Tf in T_merged. In `Transfers` I use T_merged to transfer data to T directly. In order to provide a reference for those who encounter similar problems later, I show the input sentence below:

```
[AuxVariables]
  [./T_merged]
    family = LAGRANGE
    order = FIRST
    block = 'cool clad fuel'
  [../]
[]

[AuxKernels]
  [./Ts_to_T_merged]
    type = SelfAux
    variable = T_merged
    block = 'fuel clad'
    v = Ts
  [../]
  [./Tf_to_T_merged]
    type = SelfAux
    variable = T_merged
    block = 'cool'
    v = Tf
  [../]
[]
```

By the way, will you add `MultiAppGeneralFieldTransfer` to MOOSE's Transfers system in the future? I prefer to use this object after updating MOOSE, rather than using T_merged.

**Reply by GiudGiud** [reply]
I ll work on it in the coming month

**Reply by js-jixu** [reply]
Looking forward to it!
